### ISSUE TITLE
random error in pymor cache

### ISSUE THREAD

**Issue by andreasbuhr** [pymor/pymor#117]
Hi,

I randomly get error messages like:

Traceback (most recent call last):
  File "experiment_compare_estimates.py", line 23, in <module>
    indicator=indicator
  File "/l/abuhr_01/pymor-arbilomod/arbilomod_parametric_run.py", line 240, in run_simulation
    )
  File "/l/abuhr_01/pymor-arbilomod/arbilomod_manager.py", line 475, in extend
    boundary_values = self.l.localize_vector_array(U_boundary, space['cenv'])
  File "/l/abuhr_01/pymor-for-pymor-arbilomod/src/pymor/core/cache.py", line 323, in __call__
    region.set(key, value)
  File "/l/abuhr_01/pymor-for-pymor-arbilomod/src/pymor/core/cache.py", line 181, in set
    fd = os.open(file_path, os.O_WRONLY | os.O_EXCL | os.O_CREAT)
OSError: [Errno 17] File exists: '/tmp/pymor.cache.abuhr_01/2015-05-15T21:22:44.468276.dat'

**Comment by sdrave** [comment]
Do you have multiple instances writing to the same cache at the same time? What filesystem?
Also, the traceback seems somehow incomplete or am I missing something? (Which one is the cached method?)

**Comment by andreasbuhr** [comment]
Yes, there are multiple instances running. It is on metatron. /tmp is mounted as:

/dev/mapper/ivv5_lvm-root on / type ext4 (rw,relatime,errors=remount-ro)

The traceback is complete. I am also not able to understand it. Maybe the source was changed after program start, so the line numbers are misleading. Anyhow, the reason is quite clear: There is a race condition in the "set" function (cache.py:181). 

```
        while os.path.exists(file_path):
            now = now + datetime.timedelta(microseconds=1)
            filename = now().isoformat()
            file_path = os.path.join(self.path, filename)
        fd = os.open(file_path, os.O_WRONLY | os.O_EXCL | os.O_CREAT)
```
When two processes check for the file existence at the same time, they both are informed that the file does not exist. Then, in the next step, they both try to create that file.

And I think the solution is also quite clear, as there is specialized functionality in the python standard lib for exactly this purpose:

https://docs.python.org/2/library/tempfile.html#tempfile.mkstemp

Do you share my opinion?

**Comment by sdrave** [comment]
I agree that using mkstemp is cleaner (I would say that the filename should still include the date.)

What really surprises me is that you actually hit this race condition with only a few concurrent processes. What disk cached methods do you have and how often are they called?

**Comment by andreasbuhr** [comment]
I don't know of all my cached methods, but at least I do have "solve" on the local problem in the enrichment. 
Back-of-an-envelope-calculation:
If it is called once a second, and a second has a million microseconds, you end up with a one in a million chance of having the same filename twice, for two processes. With ten processes it is something like one to 100 000. A day has 86.400 seconds, so one occurrence every day is about what to expect.

**Comment by sdrave** [comment]
Fixed in 071b3afecff0c52d5d195c4a392f2cad01a6e7c8.
